File: src/ruby/tormenta20/rules/pm_cap.rb
# frozen_string_literal: true

module Tormenta20
  module Rules
    # PM cap rules for variable-cost abilities (Tormenta20 core rules).
    #
    # Rule: the PM spent on a single use of a variable-cost ability cannot exceed
    # the character's level in the class that provides it. If the ability comes
    # from a race, origin, or general power (no class source), the cap equals the
    # total character level.
    module PmCap
      # Returns the PM cap for a variable-cost ability.
      #
      # @param source_class [String, nil] name of the class that provides the
      #   ability (e.g. "arcanista"), or nil for race/origin/general powers
      # @param class_level [Integer, nil] character's level in that class
      # @param character_level [Integer] total character level (used as fallback)
      # @return [Integer]
      def self.for_ability(source_class:, class_level:, character_level:)
        return character_level if source_class.nil? || class_level.nil?

        class_level
      end
    end
  end
end

File: src/ts/rules/pm_cap.ts
/**
 * Rules utilities for PM (Mana Points) cap calculations.
 */
export const PmCap = {
  /**
   * Returns the effective PM cap for a class ability or power.
   *
   * When a power's PM cost scales with a specific class level (e.g. Arcanista's
   * Magia Aprimorada), the cap is the character's level in that class. Otherwise
   * it falls back to the total character level.
   *
   * @param params.characterLevel - Total character level (sum of all class levels)
   * @param params.classLevel - Character's level in the class that provides the ability, or `null`
   * @param params.sourceClass - ID of the class that provides the ability, or `null`
   * @returns The PM cap as an integer
   *
   * @example
   * ```ts
   * PmCap.forAbility({ characterLevel: 10, classLevel: 6, sourceClass: "arcanista" }) // 6
   * PmCap.forAbility({ characterLevel: 10, classLevel: null, sourceClass: null })      // 10
   * ```
   */
  forAbility(params: {
    characterLevel: number
    classLevel?: number | null
    sourceClass?: string | null
  }): number {
    const { characterLevel, classLevel, sourceClass } = params
    if (sourceClass == null || classLevel == null) return characterLevel
    return classLevel
  },
}
